### PR TITLE
Extend global and txn field tests for all TEAL versions and protocols

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

* Move tests from backwardCompat_test.go to fields_test.go
* Check every field on a version before it was introduced
* Enable txna fields checks
* Fix TestGlobal to prevent invalid version specs in tests map

## Test Plan

This is tests improvement PR

### CODE REVIEW

File: data/transactions/logic/fields_test.go [hunk 1]
```diff
@@ -17,12 +17,241 @@
...
+		for v := uint64(0); v < field.version; v++ {
+			testLine(t, text, v, "...available in version...")
+		}
+		testLine(t, text, field.version, "")
+
+		ops := testProg(t, text, AssemblerMaxVersion)
+
+		// check on a version before the field version
+		preLogicVersion := field.version - 1
+		preConsensusVersion := logicVersionByProto[preLogicVersion]
```
Comment (tsachiherman) [inline data/transactions/logic/fields_test.go:135]: Since it's a unit test, get the second param from the map query and require it to be true

**Review by tsachiherman — DISMISSED** [review]
Looks great, 1 small possible improvement.

File: data/transactions/logic/fields_test.go [hunk 2]
```diff
@@ -17,12 +17,241 @@
...
+		proto := config.Consensus[preConsensusVersion]
+		if preLogicVersion < appsEnabledVersion {
+			require.False(t, proto.Application)
+		}
+		ep := defaultEvalParams(nil, nil)
```
Comment (jannotti) [inline data/transactions/logic/fields_test.go:140]: I'd rather see a loop like above, for testProg, but now you would create an `ep` using defaultEvalParamsWithVersion() to test that evaluation fails before field.version, and succeeds afterward.  I think the parsing consensus.go is a distraction - in the end your just finding the version that has the right LogicSigVersion, so you might as well just supply it directly.

File: data/transactions/logic/fields_test.go (hunk 1 above)
Comment (algorandskiy) [reply data/transactions/logic/fields_test.go:135]: done

File: data/transactions/logic/fields_test.go (hunk 2 above)
Comment (algorandskiy) [reply data/transactions/logic/fields_test.go:140]: I agree. I think I wanted to play with go ast parser. Fixed by generating fake params with 0-`LogicVersion`

File: data/transactions/logic/fields_test.go [hunk 3]
```diff
@@ -17,12 +17,177 @@
...
+		if fs.version > 1 {
+			fields = append(fields, fs)
+		}
+	}
+	require.Greater(t, len(fields), 1)
+
+	consensusParamsByLogicVersion := make(map[uint64]config.ConsensusParams, LogicVersion)
+	for v := uint64(0); v <= LogicVersion; v++ {
+		consensusParamsByLogicVersion[v] = defaultEvalProtoWithVersion(v)
+	}
```
Comment (jannotti) [inline data/transactions/logic/fields_test.go:52]: No reason for the map anymore, right?  Just invoke defaultEvalProtoWIthVersion where you would do the lookup and avoid dealing with require.True(t, ok) afterward.

**Review by jannotti — APPROVED** [review]

File: data/transactions/logic/fields_test.go (hunk 3 above)
Comment (algorandskiy) [reply data/transactions/logic/fields_test.go]: true, fixed

**Review by jannotti — APPROVED** [review]

**Pull request merged**